olp-cpp-sdk-core/src/network2/winhttp/NetworkProtocolWinHttp.cpp

```diff
@@ -0,0 +1,1063 @@
+                        sizeof(flags)))
+    handle->no_compression = true;
+
+  const std::vector<std::pair<std::string, std::string>>& extraHeaders =
+      request.GetHeaders();
+
+  std::wostringstream headerStrm;
+  _locale_t loc = _create_locale(LC_CTYPE, "C");
+  bool foundContentLength = false;
+  for (size_t i = 0; i < extraHeaders.size(); i++) {
+    std::string headerName = extraHeaders[i].first.c_str();
+    std::transform(headerName.begin(), headerName.end(), headerName.begin(),
+                   ::tolower);
+
+    if (headerName.compare("content-length") == 0) {
+      foundContentLength = true;
+    }
+
+    headerStrm << headerName.c_str();
+    headerStrm << L": ";
+    headerStrm << extraHeaders[i].second.c_str();
+    headerStrm << L"\r\n";
+  }
+
+  // Set the content-length header if it does not already exist
+  if (!foundContentLength) {
+    headerStrm << L"content-length: " << contentLength << L"\r\n";
+  }
+
+  _free_locale(loc);
+
+  if (!WinHttpAddRequestHeaders(httpRequest, headerStrm.str().c_str(),
+                                DWORD(-1), WINHTTP_ADDREQ_FLAG_ADD)) {
+    EDGE_SDK_LOG_ERROR(LOGTAG,
+                       "WinHttpAddRequestHeaders() failed " << GetLastError());
+  }
+
+  /*if (request.ModifiedSince()) {
+    FILETIME ft;
+    SYSTEMTIME st;
+    UnixTimeToFileTime(request.ModifiedSince(), &ft);
+    FileTimeToSystemTime(&ft, &st);
+    WCHAR pwszTimeStr[WINHTTP_TIME_FORMAT_BUFSIZE / sizeof(WCHAR)];
+    if (WinHttpTimeFromSystemTime(&st, pwszTimeStr)) {
+      std::wostringstream headerStrm;
+      headerStrm << L"If-Modified-Since: ";
+      headerStrm << pwszTimeStr;
+      if (!WinHttpAddRequestHeaders(httpRequest, headerStrm.str().c_str(),
+                                    DWORD(-1), WINHTTP_ADDREQ_FLAG_ADD)) {
+        EDGE_SDK_LOG_ERROR(LOGTAG,
+                           "WinHttpAddRequestHeaders(if-modified-since) failed "
+                               << GetLastError());
+      }
+    }
+  }*/
+
+  /* Send the request */
+  if (!WinHttpSendRequest(httpRequest, WINHTTP_NO_ADDITIONAL_HEADERS, 0,
+                          (LPVOID)content, contentLength, contentLength,
+                          (DWORD_PTR)handle.get())) {
+    EDGE_SDK_LOG_ERROR(LOGTAG, "WinHttpSendRequest failed " << GetLastError());
+
+    std::unique_lock<std::recursive_mutex> lock(mutex_);
+
+    http_requests_.erase(id);
+    // return NetworkProtocol::ErrorIO;
+    return NetworkStatus(ErrorCode::IO_ERROR);
+  }
+  handle->http_request = httpRequest;
+
+  // return NetworkProtocol::ErrorNone;
+  return NetworkStatus(id);
+}
+
+void NetworkProtocolWinHttp::Cancel(RequestId id) {
+  std::unique_lock<std::recursive_mutex> lock(mutex_);
+  auto it = http_requests_.find(id);
+  if (it == http_requests_.end()) {
+    return;
+  }
+
+  // Just closing the handle cancels the request
+  if (it->second->http_request) {
+    WinHttpCloseHandle(it->second->http_request);
+    it->second->http_request = NULL;
```

Comment:
dont you need to also delete it from the http_requests_ map?

Reply:
it's handled in freeHandle later (when the callback WINHTTP_CALLBACK_STATUS_HANDLE_CLOSING is called)